### PR TITLE
Add client touch capability support

### PR DESCRIPTION
**Pull request by chrisduerr**
While Smithay already supports libinput's touch events, it did not
support forwarding touch events to clients yet.

This patch implements a new `TouchHandle`, which can be acquired from
`Seat::get_touch`.

### CODE REVIEW

File: src/backend/input/mod.rs
Comment (chrisduerr) [inline src/backend/input/mod.rs:341]: I've decided to change the `TouchSlot` here because it already was creating some friction inside my compositor when handling touch events and just created more problems when it came to converting the `Option<TouchSlot>` back to a `i32` to send it to the client.

So since I'd assume nobody is using touch yet (since it doesn't work for clients), changing it now will prevent headaches in the future.

File: src/wayland/seat/mod.rs
Comment (i509VCB) [inline src/wayland/seat/mod.rs:342]: This example doesn't really explain anything more than the function docs itself.

If the seat creation are made visible it would be more jusitifed to have.

Comment (chrisduerr) [reply src/wayland/seat/mod.rs:342]: Correct, but I'd say the same applies for the existing code examples for keyboard/pointer?

> If the seat creation are made visible it would be more jusitifed to have.

What do you mean by this? Removing the `#` before the seat stuff?

**Comment by chrisduerr** [comment]
I've uploaded some videos for the people unable to test with touchscreens (these will be gone in 24h, unfortunately the videos are too big for GH):

Touch Input

Multi-touch

File: src/wayland/seat/touch.rs
Comment (PolyMeilex) [inline src/wayland/seat/touch.rs:117]: The serial can be used by the client to request some action, so the compositor needs to be aware of them. Otherwise it can't implement it's own serial checking rules. 
So we should not hide any serials from the user.

@vberger Correct me if I'm wrong

File: src/wayland/seat/touch.rs
Comment (PolyMeilex) [inline src/wayland/seat/touch.rs:124]: Same here.

File: src/wayland/seat/touch.rs
Comment (PolyMeilex) [inline src/wayland/seat/touch.rs:139]: This is missing a version check, shape is available since version 6

File: src/wayland/seat/touch.rs
Comment (PolyMeilex) [inline src/wayland/seat/touch.rs:143]: Version check as well, orientation is available since version 6.

**Review by PolyMeilex — CHANGES_REQUESTED** [review]
I have no way to test it, but overall this looks really good 👍
Just a few minor nitpicks.

File: src/wayland/seat/touch.rs
Comment (chrisduerr) [reply src/wayland/seat/touch.rs:117]: Just for my personal information: What kind of action can clients request here?

File: src/wayland/seat/touch.rs
Comment (chrisduerr) [reply src/wayland/seat/touch.rs:139]: Do you know if there's some minimum level Smithay assumes to be present? Because I noticed release is only part of version 3 but I don't think Smithay ever checks version for that?

Comment (Drakulix) [reply src/wayland/seat/touch.rs:139]: > Do you know if there's some minimum level Smithay assumes to be present? Because I noticed release is only part of version 3 but I don't think Smithay ever checks version for that?

That would be a bug then. Compositors are required to support all previous versions for stable protocols.

File: src/wayland/seat/touch.rs
Comment (PolyMeilex) [reply src/wayland/seat/touch.rs:117]: I'm not familiar with touch based clients, so I don't know what they do often, but it can basically be anything, lets say xdg-shell move/resize, or show_window_menu for example.

File: src/wayland/seat/touch.rs
Comment (chrisduerr) [reply src/wayland/seat/touch.rs:139]: I think Smithay gets around this by never actually handling release but just using automatic destructors using `assign_destructor`. So I think the existing implementations are okay? I just noticed that there weren't any version checks for release.

**Review by PolyMeilex — APPROVED** [review]
Ok, looks good to me now.

**Comment by PolyMeilex** [comment]
Actually, could we get a short changelog entry for this? Quick note about the addition and maybe mention the braking change of touch slot (although I doubt anyone was using it)

**Comment by chrisduerr** [comment]
> Actually, could we get a short changelog entry for this? Quick note about the addition and maybe mention the braking change of touch slot (although I doubt anyone was using it)

Yep, of course. Was just waiting for feedback on the general design first.

File: CHANGELOG.md
Comment (PolyMeilex) [inline CHANGELOG.md:17]: Technically, this should be under [Additions > Clients & Protocols](https://github.com/Smithay/smithay/blob/032c0de390d17219f53065bb69c99d5ef1762974/CHANGELOG.md#clients--protocols-1) as this is not "Braking Change" per se

**Review by Drakulix — APPROVED** [review]
I agree with @PolyMeilex's remarks, but other than that, this is good for merge.

**Comment by Drakulix** [comment]
@chrisduerr Could you rebase on master? I pushed some fixes to make CI on master green again and would like to verify CI works on this branch as well before merging it.

**Comment by chrisduerr** [comment]
Done.

**Comment by PolyMeilex** [comment]
Thank you!

**Pull request merged**
